### ISSUE TITLE
[VMAccessForLinux-1.4.1.0]reset ssh key in ARM mode will cause that user can login without password from any machine

### ISSUE THREAD

**Issue by yuxisun1217** [Azure/azure-linux-extensions#218]
Description of problem:
reset ssh key from Azure portal in ARM mode will cause that anyone can use this account to login without password
Only exists in ARM mode. Doesn't exist in ASM mode. 
print protect_settings in vmaccess.py:
{u'username': u'azureuser', u'password': u'', u'ssh_key': u'ssh-rsa xxx', u'reset_ssh': False}

Version-Release number of selected component:
Platform: RHEL-7
Extension: VMAccessForLinux-1.4.1.0

How reproducible:
100%

Steps to Reproduce:
1. Prepare a VM on Azure ARM mode. Username: azureuser. Authenticate with ssh key(no password)
2. Reset SSH public key of "azureuser" account from https://portal.azure.com:
    Virtual Machines -> All settings -> Reset password -> Reset SSH public key
3. Login with azureuser from any machine. If asks password, input enter directly

Actual results:
Can login use this account from any machine without password.

Expected results:
Can only login to this VM from the machine with the matched ssh key.

Logs:
/var/log/waagent:
2016/08/05 16:43:28.019544 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Expected handler state: enabled
2016/08/05 16:43:28.031937 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Decide which version to use
2016/08/05 16:43:28.113530 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Use version: 1.4.1.0
2016/08/05 16:43:28.126008 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Current handler state is: Enabled
2016/08/05 16:43:28.139402 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Update settings file: 2.settings
2016/08/05 16:43:28.151953 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Enable extension.
2016/08/05 16:43:28.163486 INFO [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Launch command:./vmaccess.py -enable
2016/08/05 16:43:28 VMAccess started to handle.
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] cwd is /var/lib/waagent/Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Change log file to /var/log/azure/Microsoft.OSTCExtensions.VMAccessForLinux/1.4.1.0/extension.log
2016/08/05 16:43:29.199824 INFO Event: name=Microsoft.OSTCExtensions.VMAccessForLinux, op=Enable, message=Launch command succeeded: ./vmaccess.py -enable

/var/log/azure/Microsoft.OSTCExtensions.VMAccessForLinux/1.4.1.0/extension.log:
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] sequence number is 2
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] setting file path is/var/lib/waagent/Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0/config/2.settings
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] JSON config: {"runtimeSettings": [{"handlerSettings": {"protectedSettings": "MIIDdAYJKoZIhvcNAQcDoIIDZTCCA2ECAQAxggFpMIIBZQIBADBNMDkxNzA1BgoJkiaJk/IsZAEZFidXaW5kb3dzIEF6dXJlIENSUCBDZXJ0aWZpY2F0ZSBHZW5lcmF0b3ICEHmWDNflMJiOSecsAJYUG10wDQYJKoZIhvcNAQEBBQAEggEAfLH6mncPst4UEWH/9WBS1Q2n+NqaHvbertgor6krtQSYki1sw+Rp8VlemYj1j5SFLumOcbKDhrkT0qhaj0tHvMwv2rYWmWHn7z42c8btG5H7kGpMnbXzkPMeDdRxAulLsp3JOGP0QUntn7EeNxMGjmui3/GE5z9/YkRyTYYLS2d8c/c8KKk5N4WIkcftOeRWP4+gD/PCPL9AdTt7MUki5p/ybdpIWr0dN/chjCcTTSDP6l/WSzevn79t3t8oQChWSpJgC3MB6BfmFtM9+qxW11MhQl1e4Qqe8HBdyLIvfFbqP41y9pOdUaeTnS8BR9lCZvWa82KLjSlmU35H6MIvNDCCAe0GCSqGSIb3DQEHATAUBggqhkiG9w0DBwQIwrvBLIp7EBmAggHI1I2XaTC2KicUoq8IVZbwWHVK14ZTvF1uxp+duhpSOu0K2ohrn/jIYxFbdbd5lHsgNYbm/j599c01Wn4uRdjj3rqkrB9UpUMDqqJ5E1mf7FAwP+yuU6D40ml6b4APj2d6lggkip4Z0hD4ZvV6xnNOueGw2K50PvatADRG8LBe8CHRN4OWi9juKulS+Nn9zw53QANIYDlEc2RWMrAUQOqbMcXxa42TmZAgCWTgMcdkRa7NlBOfOmpStpmu2U8HJUTXEqGkpqf3/J5zvV3QLhbheEGTb5TZMo67eleWt0kfEmH+4fikq9+YlAbRsYucLrfFSYnS1C9RFGD4fYWWmyD3LVVEb0yqRsuGcEcIiB2XUPYVoUbYSTZN3ya0ivxBdoJ3bw2oeSjwaAPOmN3ztyTFYGbvJanB!
FP3YYGnYBN1DTlbJrq8191dWx0jDxkfq5t1+3YpifBBzkMhbQKySPfgKS9AiggFmJByapND+AflEYu3cQFWY7ZzGo1VHP64thOQ97AZp+sKj3cZ1ExGQj2u22V+BSOu0J+ky15O3k2KQb8EY/eQNxKAAg6AiieoRqFARC+ZBju5TGNRVNUf8ukxV14Z07/2LA+TA", "publicSettings": {}, "protectedSettingsCertThumbprint": "E60DAF16AD4BEEDD25881D62ECA159431E8A114B"}}]}
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Config decoded correctly.
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] set most recent sequence number to 2
2016/08/05 16:43:28 CreateAccount: azureuser already exists. Will update password.
2016/08/05 16:43:28 Change user password
2016/08/05 16:43:28 Created user account: azureuser
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Succeeded in create the account or set the password.
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Succeeded in resetting ssh_key.
2016/08/05 16:43:28 [Microsoft.OSTCExtensions.VMAccessForLinux-1.4.1.0] Enable,success,0,Enable succeeded.

**Comment by boumenot** [comment]
Thank you for logging an issue, and making it so detailed!

The extension makes no distinction between ASM vs. ARM, so it's strange that it would behave differently.  The portal may pass different configuration in the case of ASM vs. ARM?!?

The sequence number in your extension.log is 2.  Were there any other operations you executed against VMAccessForLinux before?

Did you create your VM through the portal, or with an ARM template?  I'll try to reproduce the issue, and report back my findings.

**Comment by boumenot** [comment]
Here's my attempt to recreate and isolate the behavior you describe
using the CLI.

**Summary**: I cannot recreate the behavior by crafting various
VMAccessForLinux JSON payloads.  I will need to re-create via the
portal, and decrypt the payloads they are sending.

## CLI

1. Basic command to verify VMAccessForLinux installs, and works.  I executed check disk.

```sh
azure vm extension set chrboum-vmaccess97 chrboum97 VMAccessForLinux Microsoft.OSTCExtensions * --private-config-path check_disk.json
```

```json
{
  "check_disk": "true"
}
```

2. Reset the user's SSH key.

```sh
azure vm extension set chrboum-vmaccess97 chrboum97 VMAccessForLinux Microsoft.OSTCExtensions * --private-config-path check_disk.json
```

```json
{
    "username":"me",
    "ssh_key":"ssh-rsa ..."
}
```

The authorized_keys changed.  The supplied ssh_key was appended to the
list of authorized_keys.  I attempted to connect to the host using
ssh, and my connection was rejected due to a missing authentication
mechansim.

> Permission denied (publickey,gssapi-keyex,gssapi-with-mic).

3. Reset SSHD.

```sh
azure vm extension set chrboum-vmaccess97 chrboum97 VMAccessForLinux Microsoft.OSTCExtensions 1.4 --private-config-path reset-ssh.json
```

```json
{
  "reset_ssh": true
}
```

This command replaces /etc/ssh/sshd_config with the version that is
checked into VMAccessForLinux repository.  This is a fairly drastic
change, and resets the defaults established by the image publisher.

I attempted to connect to the host using ssh, and my connection was
rejected because I did not know the password of the user.  When I
created the VM in the portal I chose an Authentication type of **SSH
public key**.  This means password authentication is disabled in SSH,
and the password is set to a random value.  I could not connect with a
password because I did not know it.

> Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password)

The list of accepted authentication mechanism has changed.  Password
is now an accepted option wheres previously it was not.

4. Reset SSHD, password to an empty string, and SSH key.

```sh
azure vm extension set chrboum-vmaccess97 chrboum97 VMAccessForLinux Microsoft.OSTCExtensions 1.4 --private-config-path reset-ssh-password-key.json
```

```json
{
  "reset_ssh": true,
  "username": "me",
  "password": "",
  "ssh_key":"ssh-rsa ..."
}
```

This command incorporates 2 and 3, and sets the user's password to an
empty string.

I attempted to ssh to the host, and it failed.  A blank password was not accepted.

5. Reset SSHD, password to an empty string, and SSH key.

```sh
azure vm extension set chrboum-vmaccess97 chrboum97 VMAccessForLinux Microsoft.OSTCExtensions 1.4 --private-config-path reset-ssh-password-key.json
```

```json
{
  "reset_ssh": true,
  "username": "me",
  "password": "SecretPassword",
  "ssh_key":"ssh-rsa ..."
}
```

This command incorporates 2 and 3, and sets the user's password to a
value.

I attempted to ssh to the host, and it succeeded when I entered the
expected password.

**Comment by boumenot** [comment]
Here's my attempt to recreate and isolate the behavior you describe
using the Portal. 


## Portal (New - portal.azure.com)

1. I created a RHEL 7.2 image.  I chose an Authentication type of
**SSH public key**.

 1. Click on _Reset password_ under SUPPORT + TROUBLESHOOTING.
 1. Reset password or... dialog appears
  a. Mode is set to 'Reset SSH public key'
  b. User name is set to <username>
  c. SSH public key is set to <ssh-public-key>
 1. Wait for the portal to raise an alert saying, "Successfully reset SSH key"
 1. Confirm connectivity to the VM using the new key. [CONFIRMED]
 
I attempted to SSH to the host, and my access was denied.  I did not
supply a key, and I was not prompted with a password

> Permission denied (publickey,gssapi-keyex,gssapi-with-mic).

**Comment by boumenot** [comment]
@yuxisun1217 - I cannot recreate the behavior you are seeing.  Would you please review my process to see if I missed something?

**Comment by yuxisun1217** [comment]
Hi @boumenot ,
I reproduce it again and I'm sorry I forget this important step:
After reset ssh key, run "sudo systemctl sshd restart" inside the VM.
(Actually I think this might be another issue that vmaccess should restart sshd service by itself after reset ssh key?)

**Comment by yuxisun1217** [comment]
More information after reset ssh key:

1. The permission in /etc/sudoers.d/waagent becomes this: 
$ sudo cat /etc/sudoers.d/waagent
azureuser ALL = (ALL) ALL
(Before reset ssh key, it is: azureuser ALL = (ALL) NOPASSWD: ALL)

2. $ sudo cat /etc/shadow|grep azureuser
azureuser:$6$5lyO4PY1IZ$9q6qYLe5K.FXmy1L10g2KOmewZVQSolmitpDmNpWD4GiHCdzkgnWLSGKjGXugx.2kRG8961E9m7P9NCyl1ZaT/:17021:0:99999:7:::

**Comment by boumenot** [comment]
Sorry about the delay, I was on vacation, and I am just returning to this.

@yuxisun1217 - I dug into this a little bit more I *think* I understand the issue a little bit better.

If you pass a configuration with a **password** the VMAccess extension will update SSHD's configuration to ensure passwords are an accepted authentication mechanism.

```json
{
  "username":"azureuser",
  "password":"SuperSecret=="
}
```

The code makes the following changes to /etc/ssh/sshd_config.
* set _PasswordAuthentication_ to  "yes"
* set _ChallengeResponseAuthentication_ "yes"

The code fails to create a backup for /etc/ssh/sshd_config, and fails to restart SSHD, which it would need to do to ensure the change actually works for the end-user. I believe this explains why you had to restart SSHD for the change to take effect.  Does this agree with your understanding?

The modification to /etc/sudoers.d/waagent is dependent upon the configuration used with VMAccess. 
* If configuration contains password, then configuration is set to "ALL = (ALL) ALL"
* If configuration does not contain password, then configuration is set to "ALL = (ALL) NOPASSWD: ALL"

The ability to log into the VM with an empty password leads me to believe you specified a configuration like the following.  If so, this would re-create the behavior your described.

```json
{
  "username":"azureuser10",
  "password":""
}
```

The code allows users to set empty passwords. This is valid, but unintended in your case.

I think there are at least two fixes that need to be made here.
1. Ensure /etc/ssh/sshd_config is backed up when edits are made.  If the user specifies a password sshd_config is updated, but not backed up.
1. Ensure SSHD is restarted whenever edits are made.  If the user specifies a password sshd_config is updated, but not restarted automatically.

What do you think?

**Comment by yuxisun1217** [comment]
Hi @boumenot ,

Thank you very much for your research. 

I quite agree to the second fix. I think a sshd restart is necessary after resetting password because of the sshd_config change.

About the first fix, do you mean to backup the sshd_config file after resetting password? I think we can just keep it as before, because only the "PasswordAuthentication" and "ChallengeResponseAuthentication" parameters are changed. In my opinion, customers just use this feature to create(reset) their accounts, but not deal with connection issues. Besides, if we backup the sshd_config file, perhaps it is duplicate with the "reset remote access" feature. Anyway, backup sshd_config doesn't impact the reset password feature, I just thought there was no necessary to do that. What do you think about it?

About the empty password, I think if customers just use the CLI to set empty password, it's OK(Actually it is very dangerous and I think we'd better deny that). But my scenario is: customer intends to create an account and use public key to login. The UI is like this:
![screenshot from 2016-08-26 15-03-24](https://cloud.githubusercontent.com/assets/19280786/17996982/5df638b8-6b9e-11e6-9149-c77242c79d7b.png)
There's no place for customers to set password if they choose the "Reset SSH public key". So we should not regard it as an empty password.
On the other way, refer to the same feature on ASM mode, and refer to the create account process in the WALinuxAgent, I believe that the configuration file should be like this:
{
  "username":"azureuser10",
  "ssh_key":"xxxxx"
}

**Comment by boumenot** [comment]
I meant that anytime I am about to modify sshd_config, I will create a backup.  I do not want to modify the file without preserving an original copy.

Now I understand.  In the case of the portal, I expect it to do the right thing (I will verify), and not define an empty field.  I was more concerned with customers that explicitly define a JSON document and set the password to an empty value.

I agree, I do not think we should support blank passwords.  I'll create a new PR.

I believe all other issues have been resolved, and I am closing this issue.  Fixes are going out in VMAccess 1.4.2.0.
